common/src/test/java/io/netty/util/concurrent/DefaultPromiseTest.java

```diff
@@ -449,6 +453,177 @@ public void mustPassNullContextToContextListener() {
+    }
+
+    @Test
+    public void flatMapOnFailedFutureMustNotApplyMapperFunction() {
+        DefaultPromise<Integer> promise = new DefaultPromise<>(INSTANCE);
+        Exception cause = new Exception("boom");
+        promise.setFailure(cause);
+        AtomicInteger counter = new AtomicInteger();
+        assertThat(flatMap(promise, i -> {
+            counter.getAndIncrement();
+            return newSuccessfulPromise(INSTANCE, i.toString());
+        }).cause()).isSameAs(cause);
+        assertThat(counter.get()).isZero();
+    }
+
+    @Test
+    public void flatMapMustFailReturnedFutureWhenMapperFunctionThrows() {
+        DefaultPromise<Integer> promise = new DefaultPromise<>(INSTANCE);
+        RuntimeException cause = new RuntimeException("boom");
+        Future<Object> future = flatMap(promise, i -> {
+            throw cause;
+        });
+        promise.setSuccess(42);
+        assertThat(future.cause()).isSameAs(cause);
+    }
+
+    @Test
+    public void flatMapMustNotFailOriginalFutureWhenMapperFunctionThrows() {
+        DefaultPromise<Integer> promise = new DefaultPromise<>(INSTANCE);
+        flatMap(promise, i -> {
+            throw new RuntimeException("boom");
+        });
+        promise.setSuccess(42);
+        assertThat(promise.getNow()).isEqualTo(42);
+    }
+
+    @Test
+    public void cancelOnFutureFromFlatMapMustCancelOriginalFuture() {
+        DefaultPromise<Integer> promise = new DefaultPromise<>(INSTANCE);
+        Future<String> strFut = flatMap(promise, i -> newSuccessfulPromise(INSTANCE, i.toString()));
+        strFut.cancel(false);
+        assertTrue(promise.isCancelled());
+        assertTrue(strFut.isCancelled());
+    }
+
+    @Test
+    public void cancelOnOriginalFutureMustCancelFutureFromFlatMap() {
+        DefaultPromise<Integer> promise = new DefaultPromise<>(INSTANCE);
+        Future<String> strFut = flatMap(promise, i -> newSuccessfulPromise(INSTANCE, i.toString()));
+        promise.cancel(false);
+        assertTrue(promise.isCancelled());
+        assertTrue(strFut.isCancelled());
+    }
+
+    @Test
+    public void futureFromFlatMapMustNotCompleteUntilMappedFutureCompletes() throws Exception {
+        TestEventExecutor executor = new TestEventExecutor();
+        DefaultPromise<Integer> original = new DefaultPromise<>(executor);
+        CountDownLatch mappingLatchEnter = new CountDownLatch(1);
+        CountDownLatch mappingLatchExit = new CountDownLatch(1);
+        Future<String> strFut = flatMap(original, i -> {
+            return executor.submit(() -> {
+                mappingLatchEnter.countDown();
+                mappingLatchExit.await();
+                return i.toString();
+            });
+        });
+
+        executor.submit(() -> original.setSuccess(42));
+        mappingLatchEnter.await();
+        assertFalse(strFut.await(100));
+        mappingLatchExit.countDown();
+        assertThat(strFut.get(5, TimeUnit.SECONDS)).isEqualTo("42");
+    }
+
```

Comment:
Can we add a test where cancelling the `Future` returned from mapper function give to `flatMap` correctly terminates the returned `Future` from `flatMap`?